File: src/bigclaw/saved_views.py
from dataclasses import dataclass, field
from typing import Dict, List


VALID_VIEW_VISIBILITY = {"private", "team", "organization"}
VALID_DIGEST_CHANNELS = {"email", "slack", "webhook"}
VALID_DIGEST_CADENCES = {"hourly", "daily", "weekly"}


@dataclass(frozen=True)
class SavedViewFilter:
    field: str
    operator: str
    value: str

    def to_dict(self) -> Dict[str, str]:
        return {
            "field": self.field,
            "operator": self.operator,
            "value": self.value,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, object]) -> "SavedViewFilter":
        return cls(
            field=str(data["field"]),
            operator=str(data["operator"]),
            value=str(data["value"]),
        )


@dataclass
class SavedView:
    view_id: str
    name: str
    route: str
    owner: str
    visibility: str = "private"
    filters: List[SavedViewFilter] = field(default_factory=list)
    sort_by: str = ""
    pinned: bool = False
    is_default: bool = False

    @property
    def filter_count(self) -> int:
        return len(self.filters)

    def to_dict(self) -> Dict[str, object]:
        return {
            "view_id": self.view_id,
            "name": self.name,
            "route": self.route,
            "owner": self.owner,
            "visibility": self.visibility,
            "filters": [view_filter.to_dict() for view_filter in self.filters],
            "sort_by": self.sort_by,
            "pinned": self.pinned,
            "is_default": self.is_default,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, object]) -> "SavedView":
        return cls(
            view_id=str(data["view_id"]),
            name=str(data["name"]),
            route=str(data["route"]),
            owner=str(data["owner"]),
            visibility=str(data.get("visibility", "private")),
            filters=[SavedViewFilter.from_dict(item) for item in data.get("filters", [])],
            sort_by=str(data.get("sort_by", "")),
            pinned=bool(data.get("pinned", False)),
            is_default=bool(data.get("is_default", False)),
        )


@dataclass
class AlertDigestSubscription:
    subscription_id: str
    saved_view_id: str
    channel: str
    cadence: str
    recipients: List[str] = field(default_factory=list)
    include_empty_results: bool = False
    muted: bool = False

    def to_dict(self) -> Dict[str, object]:
        return {
            "subscription_id": self.subscription_id,
            "saved_view_id": self.saved_view_id,
            "channel": self.channel,
            "cadence": self.cadence,
            "recipients": list(self.recipients),
            "include_empty_results": self.include_empty_results,
            "muted": self.muted,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, object]) -> "AlertDigestSubscription":
        return cls(
            subscription_id=str(data["subscription_id"]),
            saved_view_id=str(data["saved_view_id"]),
            channel=str(data["channel"]),
            cadence=str(data["cadence"]),
            recipients=[str(recipient) for recipient in data.get("recipients", [])],
            include_empty_results=bool(data.get("include_empty_results", False)),
            muted=bool(data.get("muted", False)),
        )


@dataclass
class SavedViewCatalog:
    name: str
    version: str
    views: List[SavedView] = field(default_factory=list)
    subscriptions: List[AlertDigestSubscription] = field(default_factory=list)

    @property
    def view_index(self) -> Dict[str, SavedView]:
        return {view.view_id: view for view in self.views}

    def to_dict(self) -> Dict[str, object]:
        return {
            "name": self.name,
            "version": self.version,
            "views": [view.to_dict() for view in self.views],
            "subscriptions": [subscription.to_dict() for subscription in self.subscriptions],
        }

    @classmethod
    def from_dict(cls, data: Dict[str, object]) -> "SavedViewCatalog":
        return cls(
            name=str(data["name"]),
            version=str(data["version"]),
            views=[SavedView.from_dict(item) for item in data.get("views", [])],
            subscriptions=[
                AlertDigestSubscription.from_dict(item) for item in data.get("subscriptions", [])
            ],
        )


@dataclass
class SavedViewCatalogAudit:
    catalog_name: str
    version: str
    view_count: int
    subscription_count: int
    duplicate_view_names: Dict[str, List[str]] = field(default_factory=dict)
    invalid_visibility_views: List[str] = field(default_factory=list)
    views_missing_filters: List[str] = field(default_factory=list)
    duplicate_default_views: Dict[str, List[str]] = field(default_factory=dict)
    orphan_subscriptions: List[str] = field(default_factory=list)
    subscriptions_missing_recipients: List[str] = field(default_factory=list)
    subscriptions_with_invalid_channel: List[str] = field(default_factory=list)
    subscriptions_with_invalid_cadence: List[str] = field(default_factory=list)

    @property
    def readiness_score(self) -> float:
        if self.view_count == 0:
            return 0.0
        penalties = (
            len(self.duplicate_view_names)
            + len(self.invalid_visibility_views)
            + len(self.views_missing_filters)
            + len(self.duplicate_default_views)
            + len(self.orphan_subscriptions)
            + len(self.subscriptions_missing_recipients)
            + len(self.subscriptions_with_invalid_channel)
            + len(self.subscriptions_with_invalid_cadence)
        )
        score = max(0.0, 100 - ((penalties * 100) / self.view_count))
        return round(score, 1)

    def to_dict(self) -> Dict[str, object]:
        return {
            "catalog_name": self.catalog_name,
            "version": self.version,
            "view_count": self.view_count,
            "subscription_count": self.subscription_count,
            "duplicate_view_names": {
                key: list(values) for key, values in self.duplicate_view_names.items()
            },
            "invalid_visibility_views": list(self.invalid_visibility_views),
            "views_missing_filters": list(self.views_missing_filters),
            "duplicate_default_views": {
                key: list(values) for key, values in self.duplicate_default_views.items()
            },
            "orphan_subscriptions": list(self.orphan_subscriptions),
            "subscriptions_missing_recipients": list(self.subscriptions_missing_recipients),
            "subscriptions_with_invalid_channel": list(self.subscriptions_with_invalid_channel),
            "subscriptions_with_invalid_cadence": list(self.subscriptions_with_invalid_cadence),
        }

    @classmethod
    def from_dict(cls, data: Dict[str, object]) -> "SavedViewCatalogAudit":
        return cls(
            catalog_name=str(data["catalog_name"]),
            version=str(data["version"]),
            view_count=int(data.get("view_count", 0)),
            subscription_count=int(data.get("subscription_count", 0)),
            duplicate_view_names={
                str(key): [str(value) for value in values]
                for key, values in dict(data.get("duplicate_view_names", {})).items()
            },
            invalid_visibility_views=[
                str(name) for name in data.get("invalid_visibility_views", [])
            ],
            views_missing_filters=[str(name) for name in data.get("views_missing_filters", [])],
            duplicate_default_views={
                str(key): [str(value) for value in values]
                for key, values in dict(data.get("duplicate_default_views", {})).items()
            },
            orphan_subscriptions=[str(name) for name in data.get("orphan_subscriptions", [])],
            subscriptions_missing_recipients=[
                str(name) for name in data.get("subscriptions_missing_recipients", [])
            ],
            subscriptions_with_invalid_channel=[
                str(name) for name in data.get("subscriptions_with_invalid_channel", [])
            ],
            subscriptions_with_invalid_cadence=[
                str(name) for name in data.get("subscriptions_with_invalid_cadence", [])
            ],
        )


class SavedViewLibrary:
    def audit(self, catalog: SavedViewCatalog) -> SavedViewCatalogAudit:
        duplicate_view_names: Dict[str, List[str]] = {}
        invalid_visibility_views: List[str] = []
        views_missing_filters: List[str] = []
        duplicate_default_views: Dict[str, List[str]] = {}
        orphan_subscriptions: List[str] = []
        subscriptions_missing_recipients: List[str] = []
        subscriptions_with_invalid_channel: List[str] = []
        subscriptions_with_invalid_cadence: List[str] = []
        names_by_scope: Dict[str, List[str]] = {}
        defaults_by_scope: Dict[str, List[str]] = {}

        for view in catalog.views:
            scope = f"{view.route}:{view.owner}"
            names_by_scope.setdefault(scope, []).append(view.name)
            if view.is_default:
                defaults_by_scope.setdefault(scope, []).append(view.name)
            if view.visibility not in VALID_VIEW_VISIBILITY:
                invalid_visibility_views.append(view.name)
            if not view.filters:
                views_missing_filters.append(view.name)

        for scope, names in sorted(names_by_scope.items()):
            unique_names = sorted({name for name in names if names.count(name) > 1})
            if unique_names:
                duplicate_view_names[scope] = unique_names

        for scope, names in sorted(defaults_by_scope.items()):
            if len(names) > 1:
                duplicate_default_views[scope] = sorted(names)

        view_index = catalog.view_index
        for subscription in catalog.subscriptions:
            if subscription.saved_view_id not in view_index:
                orphan_subscriptions.append(subscription.subscription_id)
            if not subscription.recipients:
                subscriptions_missing_recipients.append(subscription.subscription_id)
            if subscription.channel not in VALID_DIGEST_CHANNELS:
                subscriptions_with_invalid_channel.append(subscription.subscription_id)
            if subscription.cadence not in VALID_DIGEST_CADENCES:
                subscriptions_with_invalid_cadence.append(subscription.subscription_id)

        return SavedViewCatalogAudit(
            catalog_name=catalog.name,
            version=catalog.version,
            view_count=len(catalog.views),
            subscription_count=len(catalog.subscriptions),
            duplicate_view_names=duplicate_view_names,
            invalid_visibility_views=sorted(invalid_visibility_views),
            views_missing_filters=sorted(views_missing_filters),
            duplicate_default_views=duplicate_default_views,
            orphan_subscriptions=sorted(orphan_subscriptions),
            subscriptions_missing_recipients=sorted(subscriptions_missing_recipients),
            subscriptions_with_invalid_channel=sorted(subscriptions_with_invalid_channel),
            subscriptions_with_invalid_cadence=sorted(subscriptions_with_invalid_cadence),
        )


def render_saved_view_report(catalog: SavedViewCatalog, audit: SavedViewCatalogAudit) -> str:
    lines = [
        "# Saved Views & Alert Digests Report",
        "",
        f"- Name: {catalog.name}",
        f"- Version: {catalog.version}",
        f"- Saved Views: {audit.view_count}",
        f"- Alert Subscriptions: {audit.subscription_count}",
        f"- Readiness Score: {audit.readiness_score:.1f}",
        "",
        "## Saved Views",
        "",
    ]

    if catalog.views:
        for view in catalog.views:
            filters = ", ".join(
                f"{view_filter.field}{view_filter.operator}{view_filter.value}"
                for view_filter in view.filters
            ) or "none"
            lines.append(
                f"- {view.name}: route={view.route} owner={view.owner} visibility={view.visibility} "
                f"filters={filters} sort={view.sort_by or 'none'} pinned={view.pinned} default={view.is_default}"
            )
    else:
        lines.append("- None")

    lines.extend(["", "## Alert Digests", ""])
    if catalog.subscriptions:
        for subscription in catalog.subscriptions:
            recipients = ", ".join(subscription.recipients) or "none"
            lines.append(
                f"- {subscription.subscription_id}: view={subscription.saved_view_id} channel={subscription.channel} "
                f"cadence={subscription.cadence} recipients={recipients} "
                f"include_empty={subscription.include_empty_results} muted={subscription.muted}"
            )
    else:
        lines.append("- None")

    lines.extend(["", "## Gaps", ""])
    duplicate_names = (
        "; ".join(f"{scope}={', '.join(names)}" for scope, names in audit.duplicate_view_names.items())
        if audit.duplicate_view_names
        else "none"
    )
    duplicate_defaults = (
        "; ".join(f"{scope}={', '.join(names)}" for scope, names in audit.duplicate_default_views.items())
        if audit.duplicate_default_views
        else "none"
    )
    lines.append(f"- Duplicate view names: {duplicate_names}")
    lines.append(
        f"- Invalid view visibility: {', '.join(audit.invalid_visibility_views) if audit.invalid_visibility_views else 'none'}"
    )
    lines.append(
        f"- Views missing filters: {', '.join(audit.views_missing_filters) if audit.views_missing_filters else 'none'}"
    )
    lines.append(f"- Duplicate default views: {duplicate_defaults}")
    lines.append(
        f"- Orphan subscriptions: {', '.join(audit.orphan_subscriptions) if audit.orphan_subscriptions else 'none'}"
    )
    lines.append(
        "- Subscriptions missing recipients: "
        f"{', '.join(audit.subscriptions_missing_recipients) if audit.subscriptions_missing_recipients else 'none'}"
    )
    lines.append(
        "- Subscriptions with invalid channel: "
        f"{', '.join(audit.subscriptions_with_invalid_channel) if audit.subscriptions_with_invalid_channel else 'none'}"
    )
    lines.append(
        "- Subscriptions with invalid cadence: "
        f"{', '.join(audit.subscriptions_with_invalid_cadence) if audit.subscriptions_with_invalid_cadence else 'none'}"
    )
    return "\n".join(lines) + "\n"

File: bigclaw-go/internal/product/saved_views.go
package product

import (
	"fmt"
	"net/url"
	"sort"
	"strings"

	"bigclaw-go/internal/domain"
)

var (
	validViewVisibility = map[string]struct{}{
		"private":      {},
		"team":         {},
		"organization": {},
	}
	validDigestChannels = map[string]struct{}{
		"email":   {},
		"slack":   {},
		"webhook": {},
	}
	validDigestCadences = map[string]struct{}{
		"hourly": {},
		"daily":  {},
		"weekly": {},
	}
)

type SavedViewFilter struct {
	Field    string `json:"field"`
	Operator string `json:"operator"`
	Value    string `json:"value"`
}

type SavedView struct {
	ViewID     string            `json:"view_id"`
	Name       string            `json:"name"`
	Route      string            `json:"route"`
	Owner      string            `json:"owner"`
	Visibility string            `json:"visibility"`
	Filters    []SavedViewFilter `json:"filters,omitempty"`
	SortBy     string            `json:"sort_by,omitempty"`
	Pinned     bool              `json:"pinned"`
	IsDefault  bool              `json:"is_default"`
}

type AlertDigestSubscription struct {
	SubscriptionID      string   `json:"subscription_id"`
	SavedViewID         string   `json:"saved_view_id"`
	Channel             string   `json:"channel"`
	Cadence             string   `json:"cadence"`
	Recipients          []string `json:"recipients,omitempty"`
	IncludeEmptyResults bool     `json:"include_empty_results"`
	Muted               bool     `json:"muted"`
}

type SavedViewCatalog struct {
	Name          string                    `json:"name"`
	Version       string                    `json:"version"`
	Views         []SavedView               `json:"views,omitempty"`
	Subscriptions []AlertDigestSubscription `json:"subscriptions,omitempty"`
}

type SavedViewCatalogAudit struct {
	CatalogName                     string              `json:"catalog_name"`
	Version                         string              `json:"version"`
	ViewCount                       int                 `json:"view_count"`
	SubscriptionCount               int                 `json:"subscription_count"`
	DuplicateViewNames              map[string][]string `json:"duplicate_view_names,omitempty"`
	InvalidVisibilityViews          []string            `json:"invalid_visibility_views,omitempty"`
	ViewsMissingFilters             []string            `json:"views_missing_filters,omitempty"`
	DuplicateDefaultViews           map[string][]string `json:"duplicate_default_views,omitempty"`
	OrphanSubscriptions             []string            `json:"orphan_subscriptions,omitempty"`
	SubscriptionsMissingRecipients  []string            `json:"subscriptions_missing_recipients,omitempty"`
	SubscriptionsWithInvalidChannel []string            `json:"subscriptions_with_invalid_channel,omitempty"`
	SubscriptionsWithInvalidCadence []string            `json:"subscriptions_with_invalid_cadence,omitempty"`
	ReadinessScore                  float64             `json:"readiness_score"`
}

func BuildSavedViewCatalog(tasks []domain.Task, actor, team, project string) SavedViewCatalog {
	actor = normalizedViewOwner(actor)
	scopeSuffix := viewScopeSuffix(team, project)
	views := []SavedView{
		{
			ViewID:     "active-runs" + scopeSuffix,
			Name:       "Active Runs",
			Route:      buildSavedViewRoute("/v2/control-center", team, project),
			Owner:      actor,
			Visibility: visibilityForScope(team, project),
			Filters: []SavedViewFilter{
				{Field: "state", Operator: "in", Value: "queued,leased,running,retrying"},
			},
			SortBy:    "priority:asc,updated_at:desc",
			Pinned:    true,
			IsDefault: true,
		},
		{
			ViewID:     "blocked-runs" + scopeSuffix,
			Name:       "Blocked Runs",
			Route:      buildSavedViewRoute("/v2/control-center", team, project),
			Owner:      actor,
			Visibility: visibilityForScope(team, project),
			Filters: []SavedViewFilter{
				{Field: "state", Operator: "eq", Value: "blocked"},
			},
			SortBy: "updated_at:desc",
			Pinned: true,
		},
		{
			ViewID:     "triage-inbox" + scopeSuffix,
			Name:       "Triage Inbox",
			Route:      buildSavedViewRoute("/v2/triage/center", team, project),
			Owner:      actor,
			Visibility: visibilityForScope(team, project),
			Filters: []SavedViewFilter{
				{Field: "severity", Operator: "in", Value: "critical,high"},
			},
			SortBy: "severity:desc,updated_at:desc",
			Pinned: true,
		},
		{
			ViewID:     "regressions" + scopeSuffix,
			Name:       "Regression Follow-up",
			Route:      buildSavedViewRoute("/v2/regression/center", team, project),
			Owner:      actor,
			Visibility: visibilityForScope(team, project),
			Filters: []SavedViewFilter{
				{Field: "regression_count", Operator: "gt", Value: "0"},
			},
			SortBy: "severity:desc,updated_at:desc",
			Pinned: true,
		},
		{
			ViewID:     "weekly-ops" + scopeSuffix,
			Name:       "Weekly Ops Review",
			Route:      buildSavedViewRoute("/v2/reports/weekly", team, project),
			Owner:      actor,
			Visibility: visibilityForScope(team, project),
			Filters: []SavedViewFilter{
				{Field: "window", Operator: "eq", Value: "7d"},
			},
			SortBy: "week_end:desc",
		},
	}
	if countPremium(tasks) > 0 {
		views = append(views, SavedView{
			ViewID:     "premium-runs" + scopeSuffix,
			Name:       "Premium Runs",
			Route:      buildSavedViewRoute("/v2/control-center", team, project),
			Owner:      actor,
			Visibility: visibilityForScope(team, project),
			Filters: []SavedViewFilter{
				{Field: "plan", Operator: "eq", Value: "premium"},
			},
			SortBy: "updated_at:desc",
		})
	}
	if countHighRisk(tasks) > 0 {
		views = append(views, SavedView{
			ViewID:     "high-risk" + scopeSuffix,
			Name:       "High Risk Review",
			Route:      buildSavedViewRoute("/v2/control-center", team, project),
			Owner:      actor,
			Visibility: visibilityForScope(team, project),
			Filters: []SavedViewFilter{
				{Field: "risk_level", Operator: "eq", Value: "high"},
			},
			SortBy: "priority:asc,updated_at:desc",
		})
	}
	sort.SliceStable(views, func(i, j int) bool { return views[i].ViewID < views[j].ViewID })

	subscriptions := []AlertDigestSubscription{
		{
			SubscriptionID:      "saved-view-daily-triage" + scopeSuffix,
			SavedViewID:         "triage-inbox" + scopeSuffix,
			Channel:             "slack",
			Cadence:             "daily",
			Recipients:          []string{actor},
			IncludeEmptyResults: false,
		},
		{
			SubscriptionID:      "saved-view-weekly-ops" + scopeSuffix,
			SavedViewID:         "weekly-ops" + scopeSuffix,
			Channel:             "email",
			Cadence:             "weekly",
			Recipients:          digestRecipients(tasks, actor),
			IncludeEmptyResults: true,
		},
	}
	sort.SliceStable(subscriptions, func(i, j int) bool { return subscriptions[i].SubscriptionID < subscriptions[j].SubscriptionID })

	return SavedViewCatalog{
		Name:          "operator-console-saved-views",
		Version:       "go-v1",
		Views:         views,
		Subscriptions: subscriptions,
	}
}

func AuditSavedViewCatalog(catalog SavedViewCatalog) SavedViewCatalogAudit {
	audit := SavedViewCatalogAudit{
		CatalogName:           catalog.Name,
		Version:               catalog.Version,
		ViewCount:             len(catalog.Views),
		SubscriptionCount:     len(catalog.Subscriptions),
		DuplicateViewNames:    map[string][]string{},
		DuplicateDefaultViews: map[string][]string{},
	}
	namesByScope := map[string][]string{}
	defaultsByScope := map[string][]string{}
	viewIndex := map[string]SavedView{}
	for _, view := range catalog.Views {
		scope := fmt.Sprintf("%s:%s", view.Route, view.Owner)
		namesByScope[scope] = append(namesByScope[scope], view.Name)
		if view.IsDefault {
			defaultsByScope[scope] = append(defaultsByScope[scope], view.Name)
		}
		if _, ok := validViewVisibility[view.Visibility]; !ok {
			audit.InvalidVisibilityViews = append(audit.InvalidVisibilityViews, view.Name)
		}
		if len(view.Filters) == 0 {
			audit.ViewsMissingFilters = append(audit.ViewsMissingFilters, view.Name)
		}
		viewIndex[view.ViewID] = view
	}
	for scope, names := range namesByScope {
		if duplicates := duplicateStrings(names); len(duplicates) > 0 {
			audit.DuplicateViewNames[scope] = duplicates
		}
	}
	for scope, names := range defaultsByScope {
		if len(names) > 1 {
			audit.DuplicateDefaultViews[scope] = append([]string(nil), names...)
			sort.Strings(audit.DuplicateDefaultViews[scope])
		}
	}
	for _, subscription := range catalog.Subscriptions {
		if _, ok := viewIndex[subscription.SavedViewID]; !ok {
			audit.OrphanSubscriptions = append(audit.OrphanSubscriptions, subscription.SubscriptionID)
		}
		if len(subscription.Recipients) == 0 {
			audit.SubscriptionsMissingRecipients = append(audit.SubscriptionsMissingRecipients, subscription.SubscriptionID)
		}
		if _, ok := validDigestChannels[subscription.Channel]; !ok {
			audit.SubscriptionsWithInvalidChannel = append(audit.SubscriptionsWithInvalidChannel, subscription.SubscriptionID)
		}
		if _, ok := validDigestCadences[subscription.Cadence]; !ok {
			audit.SubscriptionsWithInvalidCadence = append(audit.SubscriptionsWithInvalidCadence, subscription.SubscriptionID)
		}
	}
	sort.Strings(audit.InvalidVisibilityViews)
	sort.Strings(audit.ViewsMissingFilters)
	sort.Strings(audit.OrphanSubscriptions)
	sort.Strings(audit.SubscriptionsMissingRecipients)
	sort.Strings(audit.SubscriptionsWithInvalidChannel)
	sort.Strings(audit.SubscriptionsWithInvalidCadence)
	if len(audit.DuplicateViewNames) == 0 {
		audit.DuplicateViewNames = nil
	}
	if len(audit.DuplicateDefaultViews) == 0 {
		audit.DuplicateDefaultViews = nil
	}
	penalties := len(audit.InvalidVisibilityViews) + len(audit.ViewsMissingFilters) + len(audit.OrphanSubscriptions) + len(audit.SubscriptionsMissingRecipients) + len(audit.SubscriptionsWithInvalidChannel) + len(audit.SubscriptionsWithInvalidCadence)
	if catalog.Views == nil || len(catalog.Views) == 0 {
		audit.ReadinessScore = 0
		return audit
	}
	audit.ReadinessScore = round1(maxFloat(0, 100-(float64(penalties)*100/float64(len(catalog.Views)))))
	return audit
}

func RenderSavedViewReport(catalog SavedViewCatalog, audit SavedViewCatalogAudit) string {
	lines := []string{
		"# Saved Views & Alert Digests Report",
		"",
		fmt.Sprintf("- Name: %s", catalog.Name),
		fmt.Sprintf("- Version: %s", catalog.Version),
		fmt.Sprintf("- Saved Views: %d", audit.ViewCount),
		fmt.Sprintf("- Alert Subscriptions: %d", audit.SubscriptionCount),
		fmt.Sprintf("- Readiness Score: %.1f", audit.ReadinessScore),
		"",
		"## Saved Views",
		"",
	}
	if len(catalog.Views) == 0 {
		lines = append(lines, "- None")
	} else {
		for _, view := range catalog.Views {
			lines = append(lines, fmt.Sprintf("- %s: route=%s owner=%s visibility=%s filters=%s sort=%s pinned=%t default=%t",
				view.Name, view.Route, view.Owner, view.Visibility, renderSavedViewFilters(view.Filters), emptyFallback(view.SortBy, "none"), view.Pinned, view.IsDefault,
			))
		}
	}
	lines = append(lines, "", "## Alert Digests", "")
	if len(catalog.Subscriptions) == 0 {
		lines = append(lines, "- None")
	} else {
		for _, subscription := range catalog.Subscriptions {
			lines = append(lines, fmt.Sprintf("- %s: view=%s channel=%s cadence=%s recipients=%s include_empty=%t muted=%t",
				subscription.SubscriptionID, subscription.SavedViewID, subscription.Channel, subscription.Cadence, emptyFallback(strings.Join(subscription.Recipients, ", "), "none"), subscription.IncludeEmptyResults, subscription.Muted,
			))
		}
	}
	lines = append(lines, "", "## Gaps", "")
	lines = append(lines, fmt.Sprintf("- Duplicate view names: %s", renderSavedViewScopeMap(audit.DuplicateViewNames)))
	lines = append(lines, fmt.Sprintf("- Invalid view visibility: %s", emptyFallback(strings.Join(audit.InvalidVisibilityViews, ", "), "none")))
	lines = append(lines, fmt.Sprintf("- Views missing filters: %s", emptyFallback(strings.Join(audit.ViewsMissingFilters, ", "), "none")))
	lines = append(lines, fmt.Sprintf("- Duplicate default views: %s", renderSavedViewScopeMap(audit.DuplicateDefaultViews)))
	lines = append(lines, fmt.Sprintf("- Orphan subscriptions: %s", emptyFallback(strings.Join(audit.OrphanSubscriptions, ", "), "none")))
	lines = append(lines, fmt.Sprintf("- Subscriptions missing recipients: %s", emptyFallback(strings.Join(audit.SubscriptionsMissingRecipients, ", "), "none")))
	lines = append(lines, fmt.Sprintf("- Subscriptions with invalid channel: %s", emptyFallback(strings.Join(audit.SubscriptionsWithInvalidChannel, ", "), "none")))
	lines = append(lines, fmt.Sprintf("- Subscriptions with invalid cadence: %s", emptyFallback(strings.Join(audit.SubscriptionsWithInvalidCadence, ", "), "none")))
	return strings.Join(lines, "\n") + "\n"
}

func buildSavedViewRoute(base, team, project string) string {
	values := url.Values{}
	if team = strings.TrimSpace(team); team != "" {
		values.Set("team", team)
	}
	if project = strings.TrimSpace(project); project != "" {
		values.Set("project", project)
	}
	if len(values) == 0 {
		return base
	}
	return base + "?" + values.Encode()
}

func viewScopeSuffix(team, project string) string {
	parts := make([]string, 0, 2)
	if team = strings.TrimSpace(team); team != "" {
		parts = append(parts, team)
	}
	if project = strings.TrimSpace(project); project != "" {
		parts = append(parts, project)
	}
	if len(parts) == 0 {
		return ""
	}
	return "-" + strings.Join(parts, "-")
}

func visibilityForScope(team, project string) string {
	if strings.TrimSpace(team) != "" || strings.TrimSpace(project) != "" {
		return "team"
	}
	return "private"
}

func normalizedViewOwner(actor string) string {
	actor = strings.TrimSpace(actor)
	if actor == "" {
		return "viewer"
	}
	return actor
}

func digestRecipients(tasks []domain.Task, actor string) []string {
	seen := map[string]struct{}{}
	recipients := make([]string, 0, 4)
	appendRecipient := func(value string) {
		value = strings.TrimSpace(value)
		if value == "" {
			return
		}
		if _, ok := seen[value]; ok {
			return
		}
		seen[value] = struct{}{}
		recipients = append(recipients, value)
	}
	appendRecipient(actor)
	for _, task := range tasks {
		appendRecipient(task.Metadata["owner"])
		appendRecipient(task.Metadata["reviewer"])
		appendRecipient(task.Metadata["created_by"])
		if len(recipients) >= 3 {
			break
		}
	}
	if len(recipients) == 0 {
		recipients = append(recipients, "viewer")
	}
	sort.Strings(recipients)
	return recipients
}

func countPremium(tasks []domain.Task) int {
	count := 0
	for _, task := range tasks {
		if strings.EqualFold(strings.TrimSpace(task.Metadata["plan"]), "premium") {
			count++
		}
	}
	return count
}

func countHighRisk(tasks []domain.Task) int {
	count := 0
	for _, task := range tasks {
		if task.RiskLevel == domain.RiskHigh {
			count++
		}
	}
	return count
}

func duplicateStrings(values []string) []string {
	counts := map[string]int{}
	for _, value := range values {
		value = strings.TrimSpace(value)
		if value == "" {
			continue
		}
		counts[value]++
	}
	out := make([]string, 0)
	for value, count := range counts {
		if count > 1 {
			out = append(out, value)
		}
	}
	sort.Strings(out)
	return out
}

func renderSavedViewFilters(filters []SavedViewFilter) string {
	if len(filters) == 0 {
		return "none"
	}
	parts := make([]string, 0, len(filters))
	for _, filter := range filters {
		parts = append(parts, filter.Field+filter.Operator+filter.Value)
	}
	return strings.Join(parts, ", ")
}

func renderSavedViewScopeMap(values map[string][]string) string {
	if len(values) == 0 {
		return "none"
	}
	scopes := make([]string, 0, len(values))
	for scope := range values {
		scopes = append(scopes, scope)
	}
	sort.Strings(scopes)
	parts := make([]string, 0, len(scopes))
	for _, scope := range scopes {
		names := append([]string(nil), values[scope]...)
		sort.Strings(names)
		parts = append(parts, fmt.Sprintf("%s=%s", scope, strings.Join(names, ", ")))
	}
	return strings.Join(parts, "; ")
}

func emptyFallback(value, fallback string) string {
	value = strings.TrimSpace(value)
	if value == "" {
		return fallback
	}
	return value
}

func round1(value float64) float64 {
	return float64(int(value*10+0.5)) / 10
}

func maxFloat(left, right float64) float64 {
	if left > right {
		return left
	}
	return right
}
